### PR TITLE
enhancement!: CLI overhaul

### LINKED ISSUE #217: Database endpoint usage feedback
It's probably nonstandard behavior to always print feedback to the console about the currently-used database endpoint when the connection is created. I tentatively think we should be making a `logging` call instead. We could also make an optional `verbose` mode or something, too, but that would be more work.

### LINKED ISSUE #293: CLI refactor
I think the `CLI` module should basically house the bare minimum level of "standalone app" functionality. There's some stuff in there that might be reusable by another application (for example, we actually import parts of the disease CLI in therapy because it's an easy way to run the updating routines). Maybe we could move some of that up to the `etl` module or something.

Since the CLI is supposed to be a "standalone app", this could also give us some leeway to set more humane logging parameters there. Currently, DynamoDB produces around 15 lines of logs every second. This might be slowing us down, and also makes actually using the logs kind of difficult since 99.9% of the time the problem is not with some arcane botocore connection (and if it was, I'd never be able to figure it out anyway...). It'd be bad form to change the botocore/boto3 logging levels in importable parts of the normalizer, but we could probably do it within the CLI (maybe enable verbose logging with a CLI option?).

### PR DESCRIPTION
**Pull request by jsstevenson**
close #293
close #217 
progress on #298

stuff I did:

* reorganize the CLI stuff to more closely match unix tool/click patterns. Instead of a bunch of different entry points, use a common one (`gene-normalizer`) and then different actions are _commands_ (eg `gene-normalizer update`).
* For the update command, `sources` are now _arguments_, not an _option_. Since the number of sources can be 0 to n, I think it makes more sense to be used as an argument instead of a param passed to an option (avoids the quoting, spacing etc stuff). 
* add a `--silent` option to suppress feedback to console (false by default) in cases like pipelines that process stuff going to stdout. Addresses #217 with a very stringent check (value must be `True`, can't just be truthy, so no accidental silencing - hence the `noqa`).
* Some cleanup on CLI descriptions. Also add a CLI autosummary to docs instead of maintaining a separate description -- there's some fun (sort of janky) stuff done to make it look okay in both.
* Refactor a lot of the internal CLI update methods to a public module, `gene.etl.update`. That way they can easily be used programmatically (this is something that Thera-Py does with the Disease Normalizer, for example).

The updated docs page looks like this: https://gene-normalizer--299.org.readthedocs.build/en/299/managing_data/loading_and_updating_data.html

And this is what the help looks like for the `update` command:

```
Usage: gene-normalizer update [OPTIONS] [SOURCES]...

  Update provided normalizer SOURCES in the gene database.

  Valid SOURCES are "HGNC", "NCBI", and "Ensembl" (case is irrelevant).
  SOURCES are optional, but if not provided, either --all or --normalize must
  be used.

  For example, the following command will update NCBI and HGNC source records:

      $ gene-normalizer update HGNC NCBI

  To completely reload all source records and construct normalized concepts,
  use the --all and --normalize options:

      $ gene-normalizer update --all --normalize

  The Gene Normalizer will fetch the latest available data from all sources if
  local data is out-of-date. To suppress this and force usage of local files
  only, use the –use_existing flag:

      $ gene-normalizer update --all --use_existing

Options:
  --all           Update records for all sources.
  --normalize     Create normalized records.
  --db_url TEXT   URL endpoint for the application database. Can either be a
                  URL to a local DynamoDB server (e.g.
                  "http://localhost:8001") or a libpq-compliant PostgreSQL
                  connection description (e.g. "postgresql://postgres:password
                  @localhost:5432/gene_normalizer").
  --aws_instance  Use cloud DynamodDB instance.
  --use_existing  Use most recent local source data instead of fetching latest
                  version
  -s, --silent    Suppress console output.  
  --help          Show this message and exit.
```

### CODE REVIEW

**Comment by jsstevenson** [comment]
@korikuzma what would you think about shortening the options to `--all` and `--normalized`? eg

`gene-normalizer update --all` and `gene-normalizer update --normalized HGNC NCBI ensembl`

**Comment by korikuzma** [comment]
> @korikuzma what would you think about shortening the options to `--all` and `--normalized`? eg
> 
> `gene-normalizer update --all` and `gene-normalizer update --normalized HGNC NCBI ensembl`

@jsstevenson to confirm, `all` would be all sources _and_ concept group creation? I'm not sure about the name `normalized` for the sources.

**Comment by jsstevenson** [comment]
TODO:
* Circle back to kori comment
* Suppress noisy boto logs on CLI (?)

File: src/gene/cli.py [hunk 1]
```diff
@@ -2,67 +2,140 @@
...
+            try:
+                parsed_sources.append(SourceName[source.upper()])
+            except KeyError:
+                failed_source_names.append(source)
+        if len(failed_source_names) != 0:
+            click.echo(f"Error: unrecognized sources: {failed_source_names}")
+            click.echo(f"Valid source options are {list(SourceName)}")
+            click.get_current_context().exit(1)
+
+        parsed_sources = list(set(parsed_sources))
```
Comment (korikuzma) [inline src/gene/cli.py:105]: I don't think it's necessary to cast back to list. We could just have `parsed_sources` be a set in L93

```suggestion
        parsed_sources = set(parsed_sources)
```

File: src/gene/database/database.py [hunk 2]
```diff
@@ -261,7 +261,11 @@ class AwsEnvName(str, Enum):
 
 
 def confirm_aws_db_use(env_name: str) -> None:
-    """Check to ensure that AWS instance should actually be used."""
+    """Check to ensure that AWS instance should actually be used.
+
+    :param env_name: name of database environment. Must be `"Prod"`, `"Dev"`, or
```
Comment (korikuzma) [inline src/gene/database/database.py:266]: Could we use `AwsEnvName`?

File: src/gene/cli.py [hunk 3]
```diff
@@ -2,67 +2,140 @@
...
+            try:
+                parsed_sources.add(SourceName[source.upper()])
+            except KeyError:
+                failed_source_names.append(source)
+        if len(failed_source_names) != 0:
+            click.echo(f"Error: unrecognized sources: {failed_source_names}")
+            click.echo(f"Valid source options are {list(SourceName)}")
+            click.get_current_context().exit(1)
+
+        parsed_sources = set(parsed_sources)
```
Comment (korikuzma) [inline src/gene/cli.py:105]: Don't need this anymore
```suggestion
```

**Review by korikuzma — CHANGES_REQUESTED** [review]

File: src/gene/cli.py (hunk 3 above)
Comment (jsstevenson) [reply src/gene/cli.py:105]: sigh

**Review by korikuzma — APPROVED** [review]
🚀

**Pull request merged**